src/config.rs

```diff
@@ -632,4 +732,286 @@ mod tests {
         "###
         );
     }
+
+    #[test]
+    fn test_config_path_home_dir_existing() -> anyhow::Result<()> {
+        TestCase {
+            files: vec!["home/.jjconfig.toml"],
+            cfg: ConfigEnv {
+                home_dir: Some("home".into()),
+                ..Default::default()
+            },
+            want: Want::ExistingAndNew("home/.jjconfig.toml"),
+            ..Default::default()
+        }
+        .run()
+    }
```

Comment:
Might be better to use some parameterized testing library than crafting micro framework. We use `test-case` in lib crate, but I'm not sure if it's convenient here.

@martinvonz do you have any recommendation?

Reply:
Yes, `test-case` might work. I considered switching from `test-case` to `rstest` but I don't know if it's really better in any way we care about (it supports fixtures, but I don't know if that would help us anyway).

Reply:
Hm it doesn't seem to work great for nontrivial case descriptions. I *can* use test-case, basically something like

```
    #[test_case(
        TestCase {
            files: vec!["home/.jjconfig.toml", "config/jj/config.toml"],
            cfg: ConfigEnv {
                home_dir: Some("home".into()),
                config_dir: Some("config".into()),
                ..Default::default()
            },
            want: Want::Fail(|err| match err {
                ConfigError::AmbiguousSource(_, _) => Ok(()),
                _ => Err(anyhow::anyhow!(
                    "expected AmbiguousSource error, got: {err}"
                )),
            }),
            ..Default::default()
        }; "path_ambiguous")]
    #[test_case(
        TestCase {
            cfg: Default::default(),
            want: Want::None,
            ..Default::default()
        }; "path_none")]
    fn config_path(case: TestCase) -> anyhow::Result<()> {
        case.run()
    }
```

but the case contents will not be autoformatted due to the macro magic. Is it the direction you'd rather go? I do have to say I miss Go's framework.

Reply:
Maybe we can extract factory functions?
```
fn foo() -> TestCase { .. }
#[test_case(foo())]
#[test_case(bar())]
...
```
Not tried with `test_case`, but I think `rstest` does support this.

Reply:
How is that better from the current version, where there is a separate function for each case?

Reply:
I think we can inline `TestCase::run()` and rewrite/split to make it more easier to follow. I don't know if it's better to serve all parameters with a single test function, though.

And if each `test_case` is labeled, maybe we can remove some naming wrappers like `subrun()`?

Reply:
> I think we can inline TestCase::run() and rewrite/split to make it more easier to follow. 

If we choose to use `test_case`, sure. But to be clear, you do prefer `test_case` over separate tests (`cargo fmt` works only with the latter)?

> I don't know if it's better to serve all parameters with a single test function, though.

I suspect having a few independent test case formats would result in more code duplication/boilerplate. Also with one test case format it's easier to understand what is being tested, since you have to understand only one format / it's more consistent.

`subrun()` is there to distinguish between `new_config_path` and `existing_config_path` - which are exercices by each test case. So it's not redundant with the test case name.

Reply:
I prefer simple (and dumb) tests like
```rust
    #[test]
    fn test_config_path_none() {
        let config_env = set_up_files(...);
        ...
        assert_eq!(config_env.existing_config_path(), ...);
        assert_eq!(config_env.new_config_path(), ...);
        ...
    }
```

If `test_case()` helps to deduplicate test code (without sacrificing readability), that's nice.

My feeling is that the current `TestCase::run()` is too structured enough to write a test for the test runner itself.

> you do prefer test_case over separate tests (`cargo fmt` works only with the latter)?

If you mean embedding large code block in `#[test_case(...)]`, no, I wouldn't prefer that.

Reply:
It had Default solely to make TestCase Default, but given the current code there is no longer benefit to Default TestCase, so removed.